### DISCUSSION TITLE
relatableQuery dependsOn belongsTo value

### DISCUSSION THREAD

**Discussion by mrleblanc101** [laravel/nova-issues#4897] [Ideas & Feature Requests] [answered]
Hi,
I have this relationship structure.

Serie:
- id
- title

Seasons:
- id
- title
- serie_id

Episodes:
- id
- title
- season_id

Here is what I want to achieve:
![2022-10-11 15 56 25](https://user-images.githubusercontent.com/17392251/195186519-cf7d29dc-4ba8-4d19-a66e-0b75cb252bb3.gif)
I want the second BelongsTo to be filtered by the value of the first BelongsTo.

When I create an Episode, I need to select which Season the Episode belongsTo.
The problem is that many Series have "Season 1".

![Capture d’écran, le 2022-10-11 à 15 45 55](https://user-images.githubusercontent.com/17392251/195184791-b4bcc2e0-cbf8-47c7-ba7b-803dadc95536.png)

As a workaround, I added the Serie name to the title of the Season using this code:
```php
public function title() {
    return $this->serie->title . ' - ' . $this->title;
}
```
![Capture d’écran, le 2022-10-11 à 15 46 09](https://user-images.githubusercontent.com/17392251/195184807-56e99295-3ea4-4eea-ad5b-d00c50b16a49.png)

The problem is that I don't find this satisfactory.
I would love for the User to select the Serie in a computed dropdown (since the serie is a belongsThrought relationship, we don't want to save this on the Episode model).

I have a working prototype.
In my fields, I added a BelongsTo Serie field from a belongsThrough relationship.
This break when saving, so I used `->fillUsing(function() {}),` to not save this value, this seems like a hack.

Then, to filter the values of the Season dropdown, I created a `relatableSeasons`.
The function receive the value of the Serie dropdown on change and apply a where.
But since this function is called twice, the second time (for the associatable query) which doesn't have the param, it remove my filter. So I used the session to save the value for the second call.

```php
public static function relatableSeasons(NovaRequest $request, $query)
{
    if(!$request->viaResource()) {
        if($request->serie) {
            $request->session()->put('serie_id', $request->serie);
        }
        if($request->session()->get('serie_id')){
            return $query->where('serie_id', (int) $request->session()->get('serie_id'));
        }
    }
}
```

Any idea how I can have a `Grand-parent -> Parent ->Child` relationship without all these ugly hacks to filter the BelongsTo fields ?

**Comment by crynobone** [comment, marked as answer]
Implemented in Nova 4.19.0

You can now do the following:

```php
BelongsTo::make('User')
  ->dependsOn('role', function ($field, $request, $formData) {
    if (! empty($formData->role)) {
      $field->relatableQueryUsing(
        fn ($request, $query) => $query->whereIn('id', [$form->role]))
      );
    }
  }),
```
